Improve WebDriver error message

jushutch opened the pull request:
The current WebDriver error message is too vague, since it's a generic message with no specific guidance for each browser. Adding a link to the [documentation on WebDrivers](https://flutter.dev/docs/testing/integration-tests#running-in-a-browser)  to help troubleshooting. This same pattern exists elsewhere in the code and works well in directing people to the right information for resolving the error.

An example of the new message including the link:
```Unable to start WebDriver Session for Flutter for Web testing. 
Make sure you have the correct WebDriver Server running at 4444. 
Make sure the WebDriver Server matches option --browser-name. 
For more information see: https://flutter.dev/docs/testing/integration-tests#running-in-a-browser 
SocketException: OS Error: The remote computer refused the network connection.
, errno = 1225, address = localhost, port = 61383
```

This PR fixes issue #78472.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [X] All existing and new tests are passing.

jushutch committed:
Add link to documentation in error message

The current WebDriver message is incomplete. Add link to
the documentation on WebDrivers for different browsers
to help troubleshooting.

jonahwilliams commented:
It seems reasonable that this could be test exempt but I would ask in the discord according to the instructions above.

jonahwilliams commented:
@Hixie did this get marked as test exempt or not?

Hixie commented:
It did not (test-exempt PRs have a comment somewhere that starts with `test-exempt: `). I would recommend having a test, in part because hopefully a test would catch the error in the PR. :-)

jushutch committed:
Add test for documentation link

Add a simple test to verify that the WebDriver error
message includes a link to the documentation.

jushutch commented:
I just added a simple test to check that the error message contains the link that was added. However I must be missing the error in the PR that was mentioned; @Hixie would you mind elaborating to save some time? Is there more that I should be explicitly testing for in this change? Thanks for the help!

jonahwilliams commented on packages/flutter_tools/test/general.shard/drive/drive_service_test.dart:
```diff
@@ -366,6 +368,29 @@ void main() {
+      ),
+    );
+
+    try {
```
This is dangerous because if we accidentally change this to not throw a tool exit the test will still pass. Consider something like this:

```dart
expect(() => driverService.startTest(...), throwsToolExit(message: ...))
```

jushutch replied:
I agree, that's exactly what I was looking for and couldn't find, thanks!

Hixie commented on packages/flutter_tools/lib/src/drive/web_driver_service.dart:
```diff
@@ -111,6 +111,8 @@ class WebDriverService extends DriverService {
         'Make sure you have the correct WebDriver Server running at $driverPort. \n'
         'Make sure the WebDriver Server matches option --browser-name. \n'
+        'For more information see: '
+        'https://flutter.dev/docs/testing/integration-tests#running-in-a-browser \n'
```
the trailing space before the newline here is what i was referring to

jushutch replied:
Okay, thanks for pointing it out. I assumed this was the correct style based on the existing message, but let me know if you prefer otherwise and I can remove the trailing spaces for the entire message.

jonahwilliams replied:
Friendly ping @Hixie

Hixie replied:
The existing text also needs fixing, I hadn't noticed that.

There shouldn't be a trailing space before a newline.

jushutch replied:
I removed all of the trailing spaces before new lines in this message 👍

jushutch committed:
Expect ToolExit exception in test

The previous implementation was vulnerable to false
positives when a ToolExit exception isn't thrown. Change
to fail when no ToolExit is thrown.

jushutch committed:
Add type annotations to test

Specify types to meet style rules.

jonahwilliams reviewed (approved):
LGTM if lg to @Hixie

Hixie commented on packages/flutter_tools/test/general.shard/drive/drive_service_test.dart:
```diff
@@ -366,6 +367,25 @@ void main() {
+      <String, String>{},
+      PackageConfig(<Package>[Package('test', Uri.base)]),
+      browserName: 'chrome',
+    ), throwsToolExit(message: RegExp(' $link ')));
```
I would probably make this test more strict. It's not just that there's a link that matters, it's also that the text is correct around the link, right?

jushutch replied:
Good point, I made the test more strict by including the preamble 👍

jushutch committed:
Remove spaces before newlines

Remove trailing spaces before newlines to follow style
guide.

jushutch committed:
Make test for error message more strict

Test that the link in the error message includes an appropriate
preamble and follows style rules.

jonahwilliams commented:
This branch unfortunately has merge conflicts that need to be handled before we could land it.

jushutch committed:
Merge branch 'master' into 78472-webdriver-error-message

jonahwilliams commented:
I don't know why this got stuck so I'm merging it
